wandb/sdk/service/streams.py

```diff
@@ -0,0 +1,305 @@
+        action = StreamAction(action="del", stream_id=stream_id)
+        self._action_q.put(action)
+        action.wait_handled()
+
+    def drop_stream(self, stream_id: str) -> None:
+        action = StreamAction(action="drop", stream_id=stream_id)
+        self._action_q.put(action)
+        action.wait_handled()
+
+    def teardown(self, exit_code: int) -> None:
+        action = StreamAction(action="teardown", stream_id="na", data=exit_code)
+        self._action_q.put(action)
+        action.wait_handled()
+
+    def stream_names(self) -> List[str]:
+        with self._streams_lock:
+            names = list(self._streams.keys())
+            return names
+
+    def has_stream(self, stream_id: str) -> bool:
+        with self._streams_lock:
+            return stream_id in self._streams
+
+    def get_stream(self, stream_id: str) -> StreamRecord:
+        with self._streams_lock:
+            stream = self._streams[stream_id]
+            return stream
+
+    def _process_add(self, action: StreamAction) -> None:
+        stream = StreamRecord()
+        # run_id = action.stream_id  # will want to fix if a streamid != runid
+        settings = action._data
+        thread = StreamThread(
+            target=wandb.wandb_sdk.internal.internal.wandb_internal,
+            kwargs=dict(
+                settings=settings,
+                record_q=stream._record_q,
+                result_q=stream._result_q,
+                port=self._port,
+                user_pid=self._pid,
+            ),
+        )
+        stream.start_thread(thread)
+        with self._streams_lock:
+            self._streams[action._stream_id] = stream
+
+    def _process_del(self, action: StreamAction) -> None:
+        with self._streams_lock:
+            stream = self._streams.pop(action._stream_id)
+            stream.join()
+        # TODO: we assume stream has already been shutdown.  should we verify?
+
+    def _process_drop(self, action: StreamAction) -> None:
+        with self._streams_lock:
+            stream = self._streams.pop(action._stream_id)
+            stream.drop()
+            stream.join()
+
+    def _finish_all(self, streams: Dict[str, StreamRecord], exit_code: int) -> None:
+        if not streams:
+            return
+
+        for sid, stream in streams.items():
+            wandb.termlog(f"Finishing run: {sid}...")  # type: ignore
+            stream.interface.publish_exit(exit_code)
+
+        streams_to_join = []
+        while streams:
+            for sid, stream in list(streams.items()):
+                poll_exit_resp = stream.interface.communicate_poll_exit()
+                if poll_exit_resp and poll_exit_resp.done:
+                    streams.pop(sid)
+                    streams_to_join.append(stream)
+                time.sleep(0.1)
+
+        # TODO: this would be nice to do in parallel
+        for stream in streams_to_join:
+            stream.join()
+
+        wandb.termlog("Done!")  # type: ignore
```

Comment:
maybe more informative message?

Reply:
This was covered in this ticket already, just havent taken the time to do it yet:
https://wandb.atlassian.net/browse/WB-6948